src/dmd/target.d

```diff
@@ -774,6 +776,90 @@ struct Target
     }
+
+    /**
+     * Get targetInfo by key
+     * Params:
+     *  name = name of targetInfo to get
+     *  loc = location to use for error messages
+     * Returns:
+     *  Expression for the requested targetInfo
+     */
+    extern (C++) static Expression getTargetInfo(const(char)* name, const ref Loc loc)
+    {
+        int ival;
+        const(char)[] sval;
+        switch (name.toDString)
+        {
+            case "arch":
+                sval = global.params.is64bit ? "x86_64" : "x86";
+                goto LstringExpr;
+            case "system":
+                if (global.params.isWindows)
+                    sval = "windows";
+                else if (global.params.isLinux)
+                    sval = "linux";
+                else if (global.params.isOSX)
+                    sval = "darwin";
+                else if (global.params.isFreeBSD)
+                    sval = "freebsd";
+                else if (global.params.isOpenBSD)
+                    sval = "openbsd";
+                else if (global.params.isDragonFlyBSD)
+                    sval = "dragonfly";
+                else if (global.params.isSolaris)
+                    sval = "solaris";
+                else
+                    sval = "unknown";
+                goto LstringExpr;
+            case "objectFormat":
+                if (global.params.isWindows)
+                    sval = global.params.mscoff ? "coff" : "omf";
+                else if (global.params.isOSX)
+                    sval = "macho";
+                else
+                    sval = "elf";
+                goto LstringExpr;
+            case "bits":
+                ival = global.params.is64bit ? 64 : 32;
+                goto LintExpr;
+            case "pointerBits":
+                ival = global.params.isLP64 ? 64 : 32;
+                goto LintExpr;
+            case "floatAbi":
+                sval = "hard";
+                goto LstringExpr;
+            case "cppRuntimeLibrary":
+                if (global.params.isWindows)
+                {
+                    if (global.params.mscoff)
+                        sval = global.params.mscrtlib ? global.params.mscrtlib.toDString : "";
+                    else
+                        sval = "snn";
+                }
+                else
+                    sval = "";
+                goto LstringExpr;
+            default:
+                return new ErrorExp();
+        }
+    LintExpr:
+        return new IntegerExp(loc, ival, Type.tint32);
+    LstringExpr:
+        return new StringExp(loc, cast(void*)sval.ptr, sval.length);
+    }
+
+    /**
+     * Get all supported target-info keys
+     * Returns:
+     *  Array of targetInfo keys
+     */
+    extern (C++) static const(char*)* allTargetInfos(size_t* numTargetInfos)
+    {
+        __gshared immutable(char*[]) keys = [ "arch", "bits", "cppRuntimeLibrary", "floatAbi", "objectFormat", "pointerBits", "system" ];
```

Comment:
This could be an enum which can be used in the above function to avoid having to repeat this list.

Reply:
It's not clear what you have in mind... how do drive a switch by a string array?
Switched on indexed lookups and hope the cases stay in sync?

Reply:
```d
private extern(D)  enum TargetInfoKeys
{
     arch = "arch",
    //...
}
//
switch (cast(TargetInfoKeys)name.toDString) with(TargetInfoKeys)
         {
             case arch:
                //...
//
extern (C++) static (DArray!(const char)* allTargetInfos(size_t* numTargetInfos)
{
      __gshared DArray!(const char)[] keys = cast(DArray!(const char)[]) [ __traits(allMembers, TargetInfoKeys) ];
      *numTargetInfos = keys.length;
      return keys.ptr;

}
```

Reply:
Good, improvement is always welcome.

Reply:
`__gshared auto` would probably be a start.

Reply:
Huh? It is currently `__gshared immutable`...?

Reply:
Sorry I meant improvement over what I wrote, which was just `__gshared`.

Reply:
I don't like that the keys are repeated twice. I also don't like that there's no easy way to visually see all the available keys in the source code.

Reply:
I fixed it.